lib/frontend.ml

```diff
@@ -163,7 +164,12 @@ module Make(C: S.CONFIGURATION with type 'a io = 'a Lwt.t) = struct
       let pages = Io_page.pages num in
       List.iter
         (fun (gref, page) ->
-           let id = gref mod (1 lsl 16) in
+           let rec next () =
+             let id = nf.rx_id in
+             nf.rx_id <- (succ nf.rx_id) mod (1 lsl 16) ;
+             if Hashtbl.mem nf.rx_map id then next () else id
```

Comment:
This line + a hash table length of [1](https://github.com/hannesm/mirage-net-xen/blob/7fd911a564c2b145078385e6e3ac9af49f6ee9bd/lib/frontend.ml#L140) adds a `O(n)` operation in the hot-path. Can you increase the size of `rx_map` to get a better average complexity? Or use a (sparse) array? Or allocate and populate a temporary array before the `List.iter`? I don't know how bad this is, but might be useful to  check that the network performance is not completely killed by that patch (I don't have a Xen machine at hand currently unfortunately)